// @reactive-dot/core@0.0.0-snapshot.ffb961c/build/QueryBuilder.js
export default class Query {
    #instructions;
    constructor(instructions) {
        this.#instructions = instructions;
    }
    get instructions() {
        return Object.freeze(this.#instructions.slice());
    }
    fetchConstant(pallet, constant) {
        return this.#append({
            instruction: "fetch-constant",
            pallet,
            constant,
        });
    }
    readStorage(pallet, storage, args, options) {
        return this.#append({
            instruction: "read-storage",
            pallet,
            storage,
            args: options === undefined ? args : [...args, options],
        });
    }
    readStorages(pallet, storage, args, options) {
        return this.#append({
            instruction: "read-storage",
            pallet,
            storage,
            args: options === undefined ? args : args.map((args) => [...args, options]),
            multi: true,
        });
    }
    readStorageEntries(pallet, storage, args, options) {
        return this.#append({
            instruction: "read-storage-entries",
            pallet,
            storage,
            args: options === undefined ? args : [...args, options],
        });
    }
    callApi(pallet, api, args, options) {
        return this.#append({
            instruction: "call-api",
            pallet,
            api,
            args: options === undefined ? args : [...args, options],
        });
    }
    callApis(pallet, api, args, options) {
        return this.#append({
            instruction: "call-api",
            pallet,
            api,
            args: options === undefined ? args : args.map((args) => [...args, options]),
            multi: true,
        });
    }
    #append(instruction) {
        return new Query([...this.#instructions, instruction]);
    }
}
